### ISSUE TITLE
Unexpected "the parameter type `T` may not live long enough" at a point where `T` is not relevant

### ISSUE THREAD

**Issue by SamRodri** [rust-lang/rust#80675]
This code fails to compile at the `Box::new(foo)` line with ``error[E0310]: the parameter type `B` may not live long enough``.

My understanding is that at that point we have a variable of type `{impl Foo}` and `Foo: 'static`.  If we move the boxing code 
to a function (`box_foo`) the code compiles.

```rust
#![allow(unused)]

trait Foo: 'static { }

struct FooImpl;
impl Foo for FooImpl { }

trait Bar { }

struct Zing<B> { bar: B }
impl<B: Bar> Zing<B> {
    fn foo(self) -> impl Foo {
        FooImpl
    }

    fn foo_boxed(self) -> Box<dyn Foo> {
        let foo = self.foo();
        Box::new(foo)
        //box_foo(foo)
    }
}

fn box_foo(foo: impl Foo) -> Box<dyn Foo> {
    Box::new(foo)
}
```
```
error[E0310]: the parameter type `B` may not live long enough
  --> src/lib.rs:18:9
   |
11 | impl<B: Bar> Zing<B> {
   |      -- help: consider adding an explicit lifetime bound...: `B: 'static +`
...
18 |         Box::new(foo)
   |         ^^^^^^^^^^^^^ ...so that the type `impl Foo` will meet its required lifetime bounds
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=44581d550ee794c8fc8c96537977cbe8))

The compiler suggestion works, but adds an unwanted constrain on the `bar` field. The struct may want to build a `'static` value from a non-static reference.

`rustc --version --verbose`:
```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-pc-windows-msvc
release: 1.49.0
```

**Comment by jyn514** [comment]
> If we move the boxing code to a function (box_foo) the code compiles.

It compiles because you added `Foo: 'static` to the trait. If you remove that you get a similar error:

```
error[E0310]: the parameter type `impl Foo` may not live long enough
  --> src/lib.rs:24:5
   |
23 | fn box_foo(foo: impl Foo) -> Box<dyn Foo> {
   |                 -------- help: consider adding an explicit lifetime bound...: `impl Foo + 'static`
24 |     Box::new(foo)
   |     ^^^^^^^^^^^^^ ...so that the type `impl Foo` will meet its required lifetime bounds
```

Which happens because `Box<dyn Foo>` is the same as `Box<dyn Foo + 'static>`, but `impl Foo` doesn't have any implicit lifetime bound.

The real issue is that the code doesn't work even if you add `+ 'static` to `foo()`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=451e7d17141bf45d1ea4373fdca59937. That seems like a compiler bug: it knows `self.foo(): 'static`, but it doesn't seem to notice.

**Comment by SamRodri** [comment]
The issue is fixed in the latest stable (1.68), not sure when it was fixed.
